File: eims/academic_year.py
"""Academic Year defaults and validation shared by EIMS operational documents."""

import frappe
from frappe import _


CURRICULUM_GUARDED_DOCTYPES = {
	"EIMS Assessment",
	"EIMS Daily Accomplishment",
	"EIMS Student Attendance",
	"EIMS Teacher Assignment",
}


def apply_active_academic_year(doc, method=None):
	"""Default year-aware records to the active year and validate their context."""
	if not doc.meta.has_field("academic_year"):
		return

	if not doc.academic_year:
		doc.academic_year = frappe.db.get_value("EIMS Academic Year", {"is_active": 1}, "name")

	if not doc.academic_year:
		frappe.throw(_("Create and activate an Academic Year before creating academic records."))

	if not frappe.db.exists("EIMS Academic Year", doc.academic_year):
		frappe.throw(_("Academic Year {0} does not exist.").format(doc.academic_year))

	_resolve_result_enrollment(doc)
	_validate_subject_and_section_context(doc)
	_validate_curriculum_context(doc)
	_validate_student_context(doc)
	_validate_staff_context(doc)
	_validate_child_context(doc)


def _resolve_result_enrollment(doc):
	"""Results always use the student's placement for the selected year."""
	if doc.doctype != "EIMS Result" or not doc.student or not doc.academic_year:
		return

	enrollment = frappe.db.get_value(
		"EIMS Enrollment",
		{"student": doc.student, "academic_year": doc.academic_year, "status": "مسجل"},
		["stage", "section"],
		as_dict=True,
	)
	if not enrollment:
		frappe.throw(
			_("Create an Enrollment for this student and Academic Year before recording results.")
		)

	doc.stage = enrollment.stage
	doc.section = enrollment.section


def _validate_subject_and_section_context(doc):
	"""Keep direct subject, stage, and section links on the same curriculum path."""
	if doc.meta.has_field("subject") and doc.subject and doc.meta.has_field("stage"):
		subject_stage = frappe.db.get_value("EIMS Subject", doc.subject, "stage")
		if subject_stage:
			if doc.stage and doc.stage != subject_stage:
				frappe.throw(_("The selected subject does not belong to this stage."))
			doc.stage = subject_stage

	if not (doc.meta.has_field("section") and doc.section and doc.meta.has_field("stage")):
		return

	section_stage = frappe.db.get_value("EIMS Section", doc.section, "stage")
	if not section_stage:
		return
	if doc.stage and doc.stage != section_stage:
		frappe.throw(_("The selected section does not belong to this stage."))
	doc.stage = section_stage


def _validate_curriculum_context(doc):
	"""Require configured subject offerings once a year/stage curriculum exists."""
	if doc.doctype not in CURRICULUM_GUARDED_DOCTYPES or not doc.subject:
		return

	if not doc.stage:
		return

	curriculum_filters = {"academic_year": doc.academic_year, "stage": doc.stage}
	if not frappe.db.exists("EIMS Academic Curriculum", curriculum_filters):
		return

	curriculum_filters.update({"subject": doc.subject, "is_active": 1})
	if not frappe.db.exists("EIMS Academic Curriculum", curriculum_filters):
		frappe.throw(
			_("Subject {0} is not active in the Academic Year curriculum for this stage.").format(doc.subject)
		)


def _validate_child_context(doc):
	"""Validate parent-controlled links held in child tables."""
	if doc.doctype == "EIMS Assessment":
		_validate_enrolled_students(doc.students, doc.academic_year, doc.stage, doc.section, doc.subject)
	elif doc.doctype == "EIMS Student Attendance":
		_validate_enrolled_students(doc.rows, doc.academic_year, doc.stage, doc.section, doc.subject)
	elif doc.doctype == "EIMS School Activity":
		_validate_enrolled_students(doc.participants, doc.academic_year)
	elif doc.doctype == "EIMS Staff Attendance":
		for row in doc.rows or []:
			_validate_staff_membership(row.staff, doc.academic_year)
	elif doc.doctype == "EIMS Result":
		_validate_result_subjects(doc)
	elif doc.doctype == "EIMS Timetable":
		_validate_timetable_slots(doc)


def _validate_student_context(doc):
	if doc.doctype in {"EIMS Disciplinary Action", "EIMS Request"}:
		_validate_student_enrollment(doc.student, doc.academic_year)


def _validate_staff_context(doc):
	staff_field = {
		"EIMS Daily Accomplishment": "teacher",
		"EIMS School Activity": "supervisor",
		"EIMS Student Attendance": "teacher",
		"EIMS Teacher Assignment": "staff",
		"EIMS Timetable": "staff",
	}.get(doc.doctype)
	if staff_field:
		_validate_staff_membership(doc.get(staff_field), doc.academic_year)
	if doc.doctype in {"EIMS Daily Accomplishment", "EIMS Student Attendance", "EIMS Timetable"}:
		_validate_teacher_assignment(
			doc.get(staff_field), doc.academic_year, doc.stage, doc.section, doc.get("subject")
		)


def _validate_staff_membership(staff, academic_year):
	"""Enforce staff-year assignments once the year has been configured."""
	if not staff or not academic_year:
		return

	year_filters = {"academic_year": academic_year}
	if not frappe.db.exists("EIMS Staff Academic Assignment", year_filters):
		return
	if not frappe.db.exists("EIMS Staff Academic Assignment", {**year_filters, "staff": staff, "status": "Active"}):
		frappe.throw(_("Staff member {0} is not assigned to the selected Academic Year.").format(staff))


def _validate_teacher_assignment(staff, academic_year, stage, section=None, subject=None):
	"""Require operational records to match a teacher's class assignment."""
	if not staff or not academic_year or not stage:
		return

	assignments = frappe.get_all(
		"EIMS Teacher Assignment",
		filters={"staff": staff, "academic_year": academic_year, "stage": stage},
		fields=["section", "subject"],
	)
	for assignment in assignments:
		section_matches = not section or not assignment.section or assignment.section == section
		subject_matches = not subject or not assignment.subject or assignment.subject == subject
		if section_matches and subject_matches:
			return

	frappe.throw(
		_("Teacher {0} is not assigned to the selected stage, section, and subject.").format(staff)
	)


def _validate_enrolled_students(rows, academic_year, stage=None, section=None, subject=None):
	if not academic_year:
		return
	if subject:
		_validate_subject_stage(subject, stage)
		_validate_active_curriculum(academic_year, stage, subject)

	for row in rows or []:
		if not row.student:
			continue
		_validate_student_enrollment(row.student, academic_year, stage, section)


def _validate_student_enrollment(student, academic_year, stage=None, section=None):
	if not student or not academic_year:
		return
	filters = {"student": student, "academic_year": academic_year, "status": "مسجل"}
	if stage:
		filters["stage"] = stage
	if section:
		filters["section"] = section
	if frappe.db.exists("EIMS Enrollment", filters):
		return

	message = (
		"Student {0} does not have an active enrollment in the selected Academic Year and section."
		if section
		else "Student {0} does not have an active enrollment in the selected Academic Year."
	)
	frappe.throw(_(message).format(student))


def _validate_result_subjects(doc):
	for row in doc.results or []:
		if not row.subject:
			continue
		_validate_subject_stage(row.subject, doc.stage)
		_validate_active_curriculum(doc.academic_year, doc.stage, row.subject)


def _validate_timetable_slots(doc):
	for row in doc.slots or []:
		if not row.subject:
			continue

		subject_stage = _validate_subject_stage(row.subject, row.stage)
		row.stage = subject_stage
		if row.section:
			section_stage = frappe.db.get_value("EIMS Section", row.section, "stage")
			if section_stage and section_stage != subject_stage:
				frappe.throw(_("The selected section does not belong to this stage."))
		if doc.section and row.section and doc.section != row.section:
			frappe.throw(_("A timetable slot must use the timetable section."))
		_validate_active_curriculum(doc.academic_year, subject_stage, row.subject)


def _validate_subject_stage(subject, stage):
	subject_stage = frappe.db.get_value("EIMS Subject", subject, "stage")
	if subject_stage and stage and subject_stage != stage:
		frappe.throw(_("The selected subject does not belong to this stage."))
	return subject_stage


def _validate_active_curriculum(academic_year, stage, subject):
	if not academic_year or not stage:
		return

	filters = {"academic_year": academic_year, "stage": stage}
	if not frappe.db.exists("EIMS Academic Curriculum", filters):
		return

	filters.update({"subject": subject, "is_active": 1})
	if not frappe.db.exists("EIMS Academic Curriculum", filters):
		frappe.throw(
			_("Subject {0} is not active in the Academic Year curriculum for this stage.").format(subject)
		)


def ensure_staff_academic_assignment(doc, method=None):
	"""Create the first staff-year record from a staff member's hiring year."""
	if not doc.hired_year:
		return

	name = f"{doc.name}-{doc.hired_year}"
	if frappe.db.exists("EIMS Staff Academic Assignment", name):
		return

	frappe.get_doc(
		{
			"doctype": "EIMS Staff Academic Assignment",
			"staff": doc.name,
			"academic_year": doc.hired_year,
			"role": doc.role,
			"title": doc.title_en or doc.title_ar,
			"start_date": doc.hire_date,
		}
	).insert(ignore_permissions=True)


def backfill_staff_academic_assignments():
	"""Create missing staff-year records from existing staff hiring data."""
	created = 0
	for staff in frappe.get_all(
		"EIMS Staff", fields=["name", "hired_year", "role", "title_ar", "title_en", "hire_date"]
	):
		if not staff.hired_year:
			continue
		name = f"{staff.name}-{staff.hired_year}"
		if frappe.db.exists("EIMS Staff Academic Assignment", name):
			continue
		frappe.get_doc(
			{
				"doctype": "EIMS Staff Academic Assignment",
				"staff": staff.name,
				"academic_year": staff.hired_year,
				"role": staff.role,
				"title": staff.title_en or staff.title_ar,
				"start_date": staff.hire_date,
			}
		).insert(ignore_permissions=True)
		created += 1

	frappe.db.commit()
	return created


def backfill_academic_curriculum():
	"""Create year/stage/subject offerings from existing assignments and assessments."""
	created = 0
	curriculum_keys = set()

	for doctype in ("EIMS Teacher Assignment", "EIMS Assessment"):
		for row in frappe.get_all(doctype, fields=["academic_year", "stage", "subject"]):
			if row.academic_year and row.stage and row.subject:
				curriculum_keys.add((row.academic_year, row.stage, row.subject))

	for academic_year, stage, subject in curriculum_keys:
		name = f"{academic_year}-{stage}-{subject}"
		if frappe.db.exists("EIMS Academic Curriculum", name):
			continue
		frappe.get_doc(
			{
				"doctype": "EIMS Academic Curriculum",
				"academic_year": academic_year,
				"stage": stage,
				"subject": subject,
			}
		).insert(ignore_permissions=True)
		created += 1

	frappe.db.commit()
	return created

File: eims/public/js/academic_year.js
const YEAR_AWARE_DOCTYPES = [
	"EIMS Academic Curriculum",
	"EIMS Assessment",
	"EIMS Daily Accomplishment",
	"EIMS Disciplinary Action",
	"EIMS Enrollment",
	"EIMS Request",
	"EIMS Result",
	"EIMS School Activity",
	"EIMS Staff Academic Assignment",
	"EIMS Staff Attendance",
	"EIMS Stage Fee",
	"EIMS Student Attendance",
	"EIMS Student Payment",
	"EIMS Teacher Assignment",
	"EIMS Timetable",
];

const SUBJECT_DRIVEN_DOCTYPES = [
	"EIMS Academic Curriculum",
	"EIMS Assessment",
	"EIMS Daily Accomplishment",
	"EIMS Teacher Assignment",
];

const TEACHER_FIELDS = {
	"EIMS Daily Accomplishment": "teacher",
	"EIMS Student Attendance": "teacher",
	"EIMS Teacher Assignment": "staff",
	"EIMS Timetable": "staff",
};

function set_active_academic_year(frm) {
	if (!frm.is_new() || frm.doc.academic_year || frm.__eims_loading_active_year) {
		return;
	}

	frm.__eims_loading_active_year = true;
	frappe.db.get_value("EIMS Academic Year", { is_active: 1 }, "name").then(({ message }) => {
		if (message?.name && !frm.doc.academic_year) {
			frm.set_value("academic_year", message.name);
		}
	});
}

function section_filters(stage) {
	return { filters: { stage: stage || "__none__" } };
}

function subject_filters(stage, require_stage = false) {
	return stage ? { filters: { stage } } : require_stage ? { filters: { stage: "__none__" } } : {};
}

function subject_query(frm, require_stage = false) {
	return {
		query: "eims.queries.subject_query",
		filters: {
			academic_year: frm.doc.academic_year,
			stage: frm.doc.stage || (require_stage ? "__none__" : null),
		},
	};
}

function teacher_query(frm) {
	return {
		query: "eims.queries.teacher_query",
		filters: {
			academic_year: frm.doc.academic_year,
			stage: frm.doc.stage,
			section: frm.doc.section,
			subject: frm.doc.subject,
		},
	};
}

function clear_section_if_outside_stage(frm, stage) {
	if (!frm.doc.section || !stage) return;

	frappe.db.get_value("EIMS Section", frm.doc.section, "stage").then(({ message }) => {
		if (message?.stage && message.stage !== stage && frm.doc.section) {
			frm.set_value("section", null);
		}
	});
}

function clear_subject_if_outside_stage(frm, stage) {
	if (!frm.doc.subject || !stage) return;

	frappe.db.get_value("EIMS Subject", frm.doc.subject, "stage").then(({ message }) => {
		if (message?.stage && message.stage !== stage && frm.doc.subject) {
			frm.set_value("subject", null);
		}
	});
}

function sync_subject_stage(frm) {
	if (!frm.doc.subject) return;

	frappe.db.get_value("EIMS Subject", frm.doc.subject, "stage").then(({ message }) => {
		if (!message?.stage) return;
		if (frm.doc.stage !== message.stage) {
			frm.set_value("stage", message.stage);
		}
		clear_section_if_outside_stage(frm, message.stage);
	});
}

function load_result_enrollment(frm) {
	if (!frm.doc.student || !frm.doc.academic_year) return;

	frappe.call({
		method: "eims.queries.get_student_enrollment_context",
		args: { student: frm.doc.student, academic_year: frm.doc.academic_year },
	}).then(({ message }) => {
		if (!message) return;
		frm.set_value({ stage: message.stage || null, section: message.section || null });
	});
}

function set_context_queries(frm) {
	if (["EIMS Assessment", "EIMS Daily Accomplishment", "EIMS Enrollment", "EIMS Teacher Assignment"].includes(frm.doctype)) {
		frm.set_query("section", () => section_filters(frm.doc.stage));
	}

	if (frm.fields_dict.subject) {
		if (frm.doctype === "EIMS Academic Curriculum") {
			frm.set_query("subject", () => subject_filters(frm.doc.stage));
		} else {
			frm.set_query("subject", () => subject_query(frm));
		}
	}

	const teacher_field = TEACHER_FIELDS[frm.doctype];
	if (teacher_field && frm.fields_dict[teacher_field]) {
		frm.set_query(teacher_field, () => teacher_query(frm));
	}

	if (frm.doctype === "EIMS School Activity") {
		frm.set_query("supervisor", () => ({
			query: "eims.queries.staff_query",
			filters: { academic_year: frm.doc.academic_year },
		}));
		frm.set_query("student", "participants", () => ({
			query: "eims.queries.student_query",
			filters: { academic_year: frm.doc.academic_year },
		}));
	}

	if (frm.doctype === "EIMS Staff Attendance") {
		frm.set_query("staff", "rows", () => ({
			query: "eims.queries.staff_query",
			filters: { academic_year: frm.doc.academic_year },
		}));
	}

	if (frm.doctype === "EIMS Assessment") {
		frm.set_query("student", "students", () => ({
			query: "eims.queries.student_query",
			filters: {
				academic_year: frm.doc.academic_year,
				stage: frm.doc.stage,
				section: frm.doc.section,
				subject: frm.doc.subject,
			},
		}));
	}

	if (frm.doctype === "EIMS Student Attendance") {
		frm.set_query("subject", () => subject_query(frm, true));
		frm.set_query("student", "rows", () => ({
			query: "eims.queries.student_query",
			filters: {
				academic_year: frm.doc.academic_year,
				stage: frm.doc.stage,
				section: frm.doc.section,
				subject: frm.doc.subject,
			},
		}));
	}

	if (frm.doctype === "EIMS Result") {
		frm.set_query("student", () => ({
			query: "eims.queries.student_query",
			filters: { academic_year: frm.doc.academic_year },
		}));
		frm.set_query("subject", "results", () => subject_query(frm, true));
	}

	if (["EIMS Disciplinary Action", "EIMS Request", "EIMS Student Payment"].includes(frm.doctype)) {
		frm.set_query("student", () => ({
			query: "eims.queries.student_query",
			filters: { academic_year: frm.doc.academic_year },
		}));
	}

	if (frm.doctype === "EIMS Student Payment") {
		frm.set_query("enrollment", () => ({
			filters: { academic_year: frm.doc.academic_year || "__none__" },
		}));
	}

	if (frm.doctype === "EIMS Timetable") {
		frm.set_query("subject", "slots", () => subject_query(frm, true));
		frm.set_query("section", "slots", (doc, cdt, cdn) => {
			const row = locals[cdt][cdn];
			return section_filters(row.stage || frm.doc.stage);
		});
	}
}

for (const doctype of YEAR_AWARE_DOCTYPES) {
	frappe.ui.form.on(doctype, {
		setup: set_context_queries,
		onload: set_active_academic_year,
		academic_year(frm) {
			if (frm.doctype === "EIMS Result") load_result_enrollment(frm);
		},
		section(frm) {
			if (frm.doctype === "EIMS Student Attendance") {
				frappe.db.get_value("EIMS Section", frm.doc.section, "stage").then(({ message }) => {
					if (message?.stage) clear_subject_if_outside_stage(frm, message.stage);
				});
			}
			if (frm.doctype === "EIMS Timetable" && frm.doc.section) {
				frappe.db.get_value("EIMS Section", frm.doc.section, "stage").then(({ message }) => {
					if (message?.stage) frm.set_value("stage", message.stage);
				});
			}
		},
		student(frm) {
			if (frm.doctype === "EIMS Result") load_result_enrollment(frm);
		},
		stage(frm) {
			if (frm.doctype === "EIMS Enrollment") clear_section_if_outside_stage(frm, frm.doc.stage);
		},
	});
}

for (const doctype of SUBJECT_DRIVEN_DOCTYPES) {
	frappe.ui.form.on(doctype, { subject: sync_subject_stage });
}

frappe.ui.form.on("EIMS Enrollment", {
	refresh(frm) {
		if (frm.is_new() || ["منسحب", "منقول", "متخرج"].includes(frm.doc.status)) return;
		frm.add_custom_button(__("Create Payment"), () => {
			frappe.new_doc("EIMS Student Payment", { enrollment: frm.doc.name });
		}, __("Actions"));
	},
});

const REQUESTER_DOCTYPES = { "طالب": "EIMS Student", "ولي أمر": "EIMS Guardian", "معلم": "EIMS Staff" };

frappe.ui.form.on("EIMS Request", {
	raised_by_role(frm) {
		const doctype = REQUESTER_DOCTYPES[frm.doc.raised_by_role];
		frm.set_value({ raised_by_doctype: doctype || null, raised_by: null, raised_by_user: null });
	},
	raised_by(frm) {
		if (!frm.doc.raised_by || !frm.doc.raised_by_doctype) return;
		frappe.db.get_value(frm.doc.raised_by_doctype, frm.doc.raised_by, "user_id").then(({ message }) => {
			frm.set_value("raised_by_user", message?.user_id || null);
		});
	},
});
